### ISSUE TITLE
Existing routes stop working when burst creating VirtualServices that use kubernetes services for routing

### ISSUE THREAD

**Issue by CamelCaseNotation** [solo-io/gloo#2368]
https://solo-io.slack.com/archives/C9L6VPAUW/p1580870371033200

I first create a `VirtualService`, and confirm I'm able to reach it via curl.
I then run
```
watch 'curl -sq https://test-soak.platform9.horse/kplane/v1/mgmtplanes'
```
This sends a curl request every 2 seconds. The output for this curl when it's able to reach the Upstream service is `null` (doesn't matter what the output is really, but it helps that this is a small string response so it's easy to watch in a separate tmux pane)

I then run a script called `soak.sh` which ultimately creates resources in a new Namespace. Prior to creating the Namespace, I apply a `VirtualService` to my cluster in the Namespace that Gloo is running in (I modified the default namespace via the helm chart to be `gloo` rather than `gloo-system`)

This `VirtualService` has 17 routes, all in a form similar to below:
```
    - matchers:
      - prefix: /qbert/
      options:
        prefixRewrite: /
      routeAction:
        single:
          kube:
            port: 443
            ref:
              name: qbert
              namespace: du-test-soak-10
```
Notice I chose to refer to the k8s service rather than the Upstream that the Discovery component creates for k8s services due to the Discovery component causing issues similar to what this issue is for. For this issue, please keep in mind I have scaled the Discovery deployment to 0 replicas, thus disabling it.

The behavior I'm observing is the route I'm continuously curling every 2 seconds begins to return `no healthy upstream` for varying periods of time. Sometimes it's over a minute, other times it is 5 seconds, sometimes it's only 2, other times it's 22 seconds, etc.

I will try to provide a cleaned up, more concise script which the Solo team can use to reproduce this issue in the next couple weeks. This issue is to track that, as well as provide a single place to discuss what the possible causes could be.

cc @christian-posta

**Comment by christian-posta** [comment]
What version of Gloo OSS are you on so I can try reproduce?

**Comment by CamelCaseNotation** [comment]
@christian-posta I'm using 1.3.3

**Comment by CamelCaseNotation** [comment]
Here is a partial log from the Gloo Pod while the route which I'm continuously curling has been reporting "no healthy upstream" for several minutes:

```
Route Warning: InvalidDestinationWarning. Reason:
*v1.Upstream {kube-svc:du-test-soak-430-vouch-public-443 du-test-soak-430} not found gloo map[] nil {} [] 0} for resource gateway-proxy: invalid resource version gloo.gateway-proxy given , expected 1983468
event_loop.gloo
github.com/solo-io/go-utils/errutils.AggregateErrs
/go/pkg/mod/github.com/solo-io/go-utils@v0.13.1/errutils/aggregate_errs.go:19
runtime.goexit
/usr/local/go/src/runtime/asm_amd64.s:1357","stacktrace":"github.com/solo-io/gloo/projects/gloo/pkg/syncer.RunGlooWithExtensions.func7
/workspace/gloo/projects/gloo/pkg/syncer/setup_syncer.go:569"}
```

I tried to make it easier to read

**Comment by CamelCaseNotation** [comment]
Dunno if this is related, but I am seeing my Gloo Pod restart. There's no traffic being routed through it (besides my curl every 2 seconds)

Here is how much of each thing it is processing:
```
{
  "level": "info",
  "ts": 1580950663.7223766,
  "logger": "gloo.v1.event_loop.setup.v1.event_loop.envoyTranslatorSyncer",
  "caller": "syncer/envoy_translator_syncer.go:172",
  "msg": "Setting xDS Snapshot",
  "version": "1.3.3",
  "key": "gloo~gateway- proxy",
  "clusters": 4454,
  "listeners": 2,
  "routes": 2,
  "endpoints": 4454
}
```

Before the above logging statement, I see 2 logs back-to-back, with a delta that seems concerning to me given the amount of stuff that it's processing.

`1580950648.5601773` is the timestamp of an error which is thrown [here](https://github.com/solo-io/gloo/blob/10749124158ec46a19e2eafcac23ec047245055f/projects/gloo/pkg/syncer/envoy_translator_syncer.go#L141)

This is the log message below the above (and here is where it is [in the code](https://github.com/solo-io/gloo/blob/10749124158ec46a19e2eafcac23ec047245055f/projects/gloo/pkg/syncer/envoy_translator_syncer.go#L152)
```
{
  "level": "info",
  "ts": 1580950663.1094904,
  "logger": "gloo.v1.event_loop.setup.v1.event_loop.envoyTranslatorSyncer",
  "caller": "syncer/envoy_translator_syncer.go:152",
  "msg": "successfully updated EDS information for proxy gloo.gateway- proxy",
  "version": "1.3.3"
}
```

That makes me believe [this line](https://github.com/solo-io/gloo/blob/10749124158ec46a19e2eafcac23ec047245055f/projects/gloo/pkg/syncer/envoy_translator_syncer.go#L146) is taking 15 seconds to process all the stuff at the top of this comment.

**Comment by CamelCaseNotation** [comment]
Since I've turned discovery off, there aren't any `Upstream` objects in my cluster. I assume every in-memory `Upstream` object Gloo has corresponds to both a `cluster` in Envoy, as well as a `Service` in Kubernetes (or possibly `Endpoint`? I'm not sure)

I have 1525 `Services` in my cluster

**Comment by CamelCaseNotation** [comment]
This is the logs while also having just 2 VirtualServices with 1 route each

**Comment by CamelCaseNotation** [comment]
Interesting. My watch was showing the route was continuously stuck returning `no healthy upstream`

Then I saw it was hitting the service. That coincided with the Gloo Pod restarting. Again, I'm not pushing traffic, which AFAIK would only affect the gateway-proxy Pod since the control and data planes are split.

I think this may be on me since the logs of the previously running Gloo container indicates there's an issue connecting to the k8s apiserver:
```
creating base Settings resource client: list check   failed:
    Get https://10.21.0.1:443/apis/gloo.solo.io/v1/namespaces/gloo/settings:
        dial tcp 10.21.0.1:443: connect:
            connection refused
```

**Comment by christian-posta** [comment]
That's interesting, thanks for keeping us updated on this. I should have some time this weekend (I've been down with the flu this week) to try reproduce. Let us know anything else you discover that would be helpful.

**Comment by CamelCaseNotation** [comment]
Thanks @christian-posta, and I hope you're feeling better soon!!

I have an experiment which reproduces the issue. I can hop on a zoom to share live, or record it and share it with my thoughts as to what's happening. If we do it live, then I can also capture whatever logs from Gloo would be helpful.

**Comment by CamelCaseNotation** [comment]
FWIW, increasing the # of gateway-proxy Pods from 1-->3 helped with _a lot_ of the `no healthy upstream` errors I saw, but that could just be because less CURL requests were hitting a gateway-proxy Pod which had its Envoy being restarted due to new config + default round-robin behavior of k8s Services

**Comment by CamelCaseNotation** [comment]
https://www.envoyproxy.io/learn/backpressure

> Empty Cluster
What it is: There are no hosts available in a cluster. Note that this is NOT what happens when all hosts fail either health checks or outlier detection. In either of those situations, Envoy will notice the number of hosts is below the panic threshold and balance across all known hosts.

> Result: Returns an HTTP 503 with the body text “no healthy upstream”.

`no healthy upstream` is exactly what I'm seeing happen. So Gloo's reconfiguration of Envoy causes no hosts to be available in the cluster. Is this because all the envoy's are updated at the same time, resulting in them all being down at the same time?

**Comment by yuval-k** [comment]
i think this a bug that we fixed this week; i will confirm and update the status here

**Comment by yuval-k** [comment]
@CamelCaseNotation - can you try the latest version of gloo? i think the 503 might have already been fixed in it. one thing i'm trying to understand is where the InvalidDestinationWarning comes from.. have you seen this warning also appear on the virtual service \ proxy crd?

if the latest version does not solve, we would love to jump on a zoom call so we can figure out the root cause here.

**Comment by CamelCaseNotation** [comment]
Hey @yuval-k what's the latest version now? I'll definitely give it a whirl. Could you point me to the commit which you think fixes this issue? Thanks!

**Comment by CamelCaseNotation** [comment]
I'll go ahead and close this since a fix for it has landed in 1.2.17 and 1.3.4 🎉
